pkg/api/types.go

```diff
@@ -58,3 +61,32 @@ type NodeResourceUtilizationThresholds struct {
 	TargetThresholds ResourceThresholds
 	NumberOfNodes    int
 }
+
+type NamespacedTopologySpreadConstraint struct {
+	Namespace                 string
+	TopologySpreadConstraints []TopologySpreadConstraint
+}
+
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree to which pods may be unevenly distributed.
+	// It's the maximum permitted difference between the number of matching pods in
+	// any two topology domains of a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// are "spread" such that zone1 and zone2 each have one pod, but not zone3.
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// scheduling it onto zone1(zone2) would make the ActualSkew(2) violate MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required value. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
```

Comment:
@Huang-Wei - I believe this is inline with what you're proposing...

Reply:
Yes.

@krmayankk if the upstream API is available, I think you can simply vendor that?

Reply:
Currently the descheduler is looking at TopologySpreadConstraint's as defined in DeschedulerPolicy. Later these constraints will come directly from the Pod itself. So there is nothing to vendor, just the source of the TopologySpreadConstraint will change @Huang-Wei